src/ModelContextProtocol/Client/McpClientExtensions.cs

```diff
@@ -409,17 +442,28 @@ public static Task UnsubscribeFromResourceAsync(this IMcpClient client, string u
     /// <param name="client">The client.</param>
     /// <param name="toolName">The name of the tool to call.</param>
     /// <param name="arguments">Optional arguments for the tool.</param>
+    /// <param name="serializerOptions">The serialization options governing argument serialization.</param>
     /// <param name="cancellationToken">A token to cancel the operation.</param>
     /// <returns>A task containing the tool's response.</returns>
     public static Task<CallToolResponse> CallToolAsync(
-        this IMcpClient client, string toolName, IReadOnlyDictionary<string, object?>? arguments = null, CancellationToken cancellationToken = default)
+        this IMcpClient client,
+        string toolName,
+        IReadOnlyDictionary<string, object?>? arguments = null,
+        JsonSerializerOptions? serializerOptions = null,
+        CancellationToken cancellationToken = default)
     {
         Throw.IfNull(client);
         Throw.IfNull(toolName);
-
-        return client.SendRequestAsync<CallToolResponse>(
-            CreateRequest(RequestMethods.ToolsCall, CreateParametersDictionary(toolName, arguments)),
-            cancellationToken);
+        serializerOptions ??= McpJsonUtilities.DefaultOptions;
+        McpJsonUtilities.ValidateSerializerOptions(serializerOptions);
+        var parametersTypeInfo = serializerOptions.GetTypeInfo<IReadOnlyDictionary<string, object?>>();
```

Comment:
Do we have an end-to-end test verifying that I could define `[JsonPolymorphic]/`[JsonDerivedType]` parameter or result types and pass them through a call to `CallToolAsync` and handle it using a `[McpToolType`]? Will that all work and be AOT/trim friendly while using `JsonTypeInfo<IReadOnlyDictionary<string, object?>>`?

Reply:
Cf. https://github.com/modelcontextprotocol/csharp-sdk/pull/182#discussion_r2025194422